### ISSUE TITLE
Manually added Visual Editor XML files to "visualeditor-src" folder should appear in source folder "src"

### ISSUE THREAD

**Issue by piotrzarzycki21** [prominic/Moonshine-IDE#396]
Manually added Visual Editor XML files to `visualeditor-src` folder should appear in source folder (`src`) after refresh.

**Following scenario should be covered as part of that issu**e:
1. Copy `file.xml` (Visual Editor XML file) to `visualeditor-src\main\webapp\` folder
2. Right click on the project or some folder in `src` and choose **Refresh**

**Results:**
All new files in `visualeditor-src\main\webapp\` folder should be created created in Visual Editor source files and placed into appropriate folder.

**Comment by piotrzarzycki21** [comment]
I have spent solid amount of time to analyze how refreshing in project tree is working. I'm going to create plugin class called `RefreshProjectTreePlugin` which will simply respond for any kind of requests for refreshing. 

All logic responsible for refresh I will put there to have clarity what actually is happen. Then in the next step I will make magic which allows refresh Visual Editor stuff. Everything will be in one place and hopefully a bit simpler. 

@rat-moonshine if you have some thoughts on that let me know.

**Comment by piotrzarzycki21** [comment]
The other option is that I will create plugin only for refreshing visualeditor-src. In general that part will have following steps.
1. If project is PrimeFaces visual editor - I'm getting files and folders from `visualeditor-src` folder
2. I'm taking source folder and creating all the files which do not exists there, but they are in `visualeditor-src`. Of course that part including some kind of validation whether files are ok.
3. Once creation is finished I'm launching standard refresh for project tree

Thoughts ?

**Comment by rat-moonshine** [comment]
I'm not sure about how the magic you're envisioning, if the outcome is good enough, then the plan is good enough too! If you wants a plugin class only for refreshing Tree component, perhaps you wants to consider that by doing some util class instead of plugin, too. 

During my take with the tree component, I have seen the Tree component is full of bug. (As we have seen in case of `SystemTree`, previously). Thus, we always required some added effort to make everything rendered expectedly after an event - be it creating an item deep inside somewhere and also unfolding all the parent folders to have the newly created item selected and display, finding a wrapper item in-depth in collapsed tree, etc. Current `TreeView` component is full of junk codes (or we can say hacky ways etc.), if you can clean them up by moving to a separate location, I'm fine with that.

**Comment by rat-moonshine** [comment]
Since this current requested feature is mainly valid for Visual Editor project (other projects anyway shows up all the files inside a folder when a context-menu _refresh_ action does), so perhaps you want to concentrate on pointing if an XML is of Visual Editor type, after having a refresh to the folder. 

So in that case, your second comments/plan sounds much more direct to the intention.

**Comment by piotrzarzycki21** [comment]
I decided to go with the second option. I'm in the last step where I need to create new file. I would like to ask @rat-moonshine how to do that ? 
I would like to create file from code using Visual Editor template.

**Comment by rat-moonshine** [comment]
We can discuss this on skype next time you'll available, Piotr.

**Comment by piotrzarzycki21** [comment]
I'm available now and will be on the meeting if you prefer later on.

**Comment by rat-moonshine** [comment]
We require to hook [this](https://github.com/prominic/Moonshine-IDE/blob/features/issue_396_manually_added_visual_editor_xml_files/ide/MoonshineSharedCore/src/actionScripts/plugin/templating/TemplatingPlugin.as#L1169) method in `TemplatingPlugin` to automate the template file creation job. Until now, the method calls by the file creation dialog through attaching following event:

```
protected function openVisualEditorComponentTypeChoose(event:Event):void
{
            if (!newVisualEditorFilePopup)
            {
                ...
                newVisualEditorFilePopup.addEventListener(NewFileEvent.EVENT_NEW_FILE, onVisualEditorFileCreateRequest);
                …
            }
}
```

This listener active till the new file creation dialog exists, therefore it removes. To have listener method triggers anytime, we need to have an event at global level, i.e. `dispatcher.addEventListener(…)`.

But I don't wanted to move `NewFileEvent.EVENT_NEW_FILE` at global level, as the handler methods changes based on different file type creation. So I have created a new event only for Visual Editor file type - `NewFileEvent.EVENT_NEW_VISUAL_EDITOR_FILE`. While removing the existing `NewFileEvent.EVENT_NEW_FILE` from [openVisualEditorComponentTypeChoose()](https://github.com/prominic/Moonshine-IDE/blob/features/issue_396_manually_added_visual_editor_xml_files/ide/MoonshineSharedCore/src/actionScripts/plugin/templating/TemplatingPlugin.as#L915), `NewFileEvent.EVENT_NEW_VISUAL_EDITOR_FILE` event in [activate()](https://github.com/prominic/Moonshine-IDE/blob/features/issue_396_manually_added_visual_editor_xml_files/ide/MoonshineSharedCore/src/actionScripts/plugin/templating/TemplatingPlugin.as#L119) of TemplatingPlugin. 

We also required to update submission event type because of above changes, in `NewVisualEditorFilePopup.mxml`, overriding [onCreateButton()](https://github.com/prominic/Moonshine-IDE/blob/features/issue_396_manually_added_visual_editor_xml_files/ide/MoonshineSharedCore/src/components/popup/newFile/NewVisualEditorFilePopup.mxml#L34).

Next changes are to [createNewVisualEditorFiles()](https://github.com/prominic/Moonshine-IDE/blob/features/issue_396_manually_added_visual_editor_xml_files/ide/MoonshineSharedCore/src/visualEditor/plugin/VisualEditorRefreshFilesPlugin.as#L69) in `VisualEditorRefreshFilesPlugin`, and dispatch the new `NewFileEvent.EVENT_NEW_VISUAL_EDITOR_FILE` inside the loop.

(Note: You might not require these destination stuff at all at [here](https://github.com/prominic/Moonshine-IDE/blob/features/issue_396_manually_added_visual_editor_xml_files/ide/MoonshineSharedCore/src/visualEditor/plugin/VisualEditorRefreshFilesPlugin.as#L129))

Please, check.

**Comment by piotrzarzycki21** [comment]
I think I have finished with this task. Here is the rules which allows use this feature.

1) File copied to `visualeditor-src` folder should be XML, currently we are using `.xml` extension but in the future this probably will be something different. 
2) Root content for manually added file should be `RootDiv`. `MainApplication` as root won't be discovered. 
3) File which appear in sources won't be fully translated until we open it. First open triggers translation into code. In the other words if you copy 5 files and you won't open any of them in Visual Editor they will have empty `Div` after export to PrimeFaces.

**Comment by piotrzarzycki21** [comment]
Reopen with @JoelProminic comment here https://github.com/prominic/Moonshine-IDE/issues/392#issuecomment-427099010

**Comment by atinprominic** [comment]
@piotrzarzycki21 While testing I found that the process (in step 3) which triggers the translation of code after opening the file have some issue for few of the files.

Please check the xml file named - "Cust UserDoc Form-DXL.form.xml" which is converted using the Conversion tool. When I open this file in moonshine, it shouws blank white page. I have sent this file as part of my test result of Moonshine conversion.

**Comment by piotrzarzycki21** [comment]
@atinprominic this is completely unrelated to that issue https://github.com/prominic/Moonshine-IDE/issues/396#issuecomment-427749570. However I looked into that and see that in provided XML there are 3 PanelGrid. Third one in 9th row has only two columns instead 3 - That's why PanelGrid doesn't appear. If you get that file from converter - please report that to Andrii. You can show this comment.

**Comment by atinprominic** [comment]
@piotrzarzycki21 Thanks for the update. I have already given the test results to Andrii.

Let me ask his specifically for this scenario again.

**Comment by atinprominic** [comment]
@piotrzarzycki21 I tested considering your response to @JoelProminic and found one thing. I have recorded the video attached 

to explain. Its creating duplicate files with the same name in the project root folder. I think its unexpected behaviour.

![041018](https://user-images.githubusercontent.com/34882311/46606597-64c52f80-cb1b-11e8-92f6-9e85b2b71c43.gif)

**Comment by piotrzarzycki21** [comment]
@atinprominic Please make new tests, cause I just pushed some changes.

**Comment by piotrzarzycki21** [comment]
@atinprominic I noticed that you are copying file which is MainApplication - remember about following rules in this comments -> https://github.com/prominic/Moonshine-IDE/issues/396#issuecomment-427073956

One more:
- Files need to be copied to source folder - which is `main\webapp\`
- You can copy file into some sub folder, but you need to refresh it directly. I didn't implement nested refreshing. Ex.  `main\webapp\myview` -> You will have to right click on `myview` and refresh it.

**Comment by atinprominic** [comment]
@piotrzarzycki21 Its working fine now. If you have any specific test scenario then please let me know

**Comment by piotrzarzycki21** [comment]
I'm closing that one if something new pop up I will open separate issue.
